Memcpy optimization

rafaelsamenezes opened the pull request:
This pull request introduces an intrinsic implementation for the `memcpy` function in the symbolic execution engine, enabling more precise modeling and verification of memory copy operations in C programs.  Currently it only works for primitives, and it has an extra preprocessing cost (which makes it slower for small memcpy and faster for big memcpy). 

30s run:
```
Statistics:          33569 Files
  correct:           17695
    correct true:    10385
    correct false:    7310
  incorrect:            22
    incorrect true:     11
    incorrect false:    11
  unknown:           15852
  Score:             27552 (max: 55885)
```

This is a draft as I am doing another 30s run with this branch and master just to be extra sure.

rafaelsamenezes committed:
[memcpy] added intrinsic memcpy

rafaelsamenezes committed:
[memcpy] optimization for trivial case

rafaelsamenezes committed:
[memcpy] added test cases

rafaelsamenezes committed:
[memcpy] support for primitive memcpy between same type

rafaelsamenezes committed:
[memcpy] added null pointer check

rafaelsamenezes committed:
Apply automatic changes

rafaelsamenezes committed:
improve memcpy builtin

Refactored the intrinsic_memcpy function to simplify and streamline
its implementation. Removed unnecessary comments and refactored the
logic for better readability and maintainability. Ensured that all
code paths are covered and simplified the expression evaluation
process within memcpy operations.

rafaelsamenezes committed:
Apply automatic changes

rafaelsamenezes committed:
Update error messages for dereference failures

rafaelsamenezes committed:
Refactor and enhance the intrinsic_memcpy function

lucasccordeiro commented on src/goto-symex/builtin_functions.cpp:
```diff
@@ -1673,6 +1673,367 @@ static inline expr2tc gen_value_by_byte(
+    is_array_type(src->type) || is_array_type(dst->type) ||
+    is_struct_type(dst->type) || is_union_type(dst->type) ||
+    is_struct_type(src->type) || is_union_type(src->type))
+    return expr2tc();
```
Can you document that this is a limitation? It will be addressed in future PRs!

rafaelsamenezes replied:
Added a log_debug message with a todo.

lucasccordeiro commented on src/goto-symex/builtin_functions.cpp:
```diff
@@ -1673,6 +1673,367 @@ static inline expr2tc gen_value_by_byte(
 }
 
+// Computes the equivalent object value when considering a memcpy operation on it
+static inline expr2tc gen_byte_memcpy(
```
Shall we create unit tests for this method? The logic itself is intricate and might be prone to off-by-one or bit-shifting errors :-)

rafaelsamenezes committed:
Removed unused variables in `builtin_functions.cpp`

Removed commented-out variable declarations for `item_object` and
`item_offset`.

rafaelsamenezes committed:
Added todo message for non-primitive noncpy

intrigus-lgtm commented on regression/cstd/intrinsic_memcpy_bug4/main.c:
```diff
@@ -0,0 +1,14 @@
+  char *ptr2 = &b;
+  memcpy(ptr1+1, ptr2+1, sizeof(char));
+
+  assert(a == b);
```
What is this testing?

intrigus-lgtm commented on src/c2goto/library/string.c:
```diff
@@ -269,7 +269,7 @@ __ESBMC_HIDE:;
 }
 
-void *memcpy(void *dst, const void *src, size_t n)
+void *__memcpy_impl(void *dst, const void *src, unsigned long long n)
```
```suggestion
void *__memcpy_impl(void *dst, const void *src, size_t n)
```

intrigus-lgtm reviewed (changes requested):
Initial review.

intrigus-lgtm commented on src/goto-symex/goto_symex.h:
```diff
@@ -466,6 +466,19 @@ class goto_symext
     const code_function_call2t &func_call);
 
+  /**
+   * @brief Intrinsic call for C memset function call
```
Refers to memset instead of memcpy.

rafaelsamenezes committed:
Improve gen_byte_memcpy function documentation and code formatting

Added detailed documentation to the `gen_byte_memcpy` function,
including parameter descriptions and usage examples. Also, refactored
some minor code formatting for better readability.

rafaelsamenezes committed:
Moved `gen_byte_memcpy` to `goto_symex_utils` namespace

This will allow unit testing

rafaelsamenezes committed:
Apply automatic changes

rafaelsamenezes committed:
added unit tests for the intrinsic functions.

rafaelsamenezes committed:
Apply automatic changes

rafaelsamenezes commented on src/clang-c-frontend/clang_c_language.cpp:
```diff
@@ -392,7 +392,8 @@ _Bool __ESBMC_is_little_endian();
 int __ESBMC_rounding_mode = 0;
 
 void *__ESBMC_memset(void *, int, unsigned int);
```
Implemented as suggested

rafaelsamenezes committed:
Update memcpy and memset function signatures

Changed the parameter types of __ESBMC_memcpy and __ESBMC_memset from
unsigned long long to size_t for better compatibility with standard C
library functions.

rafaelsamenezes committed:
fixed intrinsic utils test compilation

Refactored the CMakeLists.txt to enable the intrinsic-utils-test unit
test. Updated the intrinsic_utils.test.cpp file by moving the mode_table
declaration outside of the SCENARIO function and adjusting the
argument order in gen_byte_memcpy to match the expected parameters.

rafaelsamenezes committed:
Apply automatic changes

rafaelsamenezes committed:
Fix memcpy type handling

Refactored the type handling in `gen_byte_memcpy` to use the source
type for generating masks instead of the destination type.
Additionally, updated the test cases to ensure they include the
necessary header file and correct function signatures.

rafaelsamenezes committed:
Apply automatic changes

rafaelsamenezes committed:
Merge branch 'master' into memcpy-optimization

rafaelsamenezes committed:
Disable intrinsic-utils-test temporarily

Currently, the CMake configuration for this test is causing issues due
to dependency problems. It does not make sense for this test to depend
on solvers at present. We will reactivate it when the dependency
issues are resolved.

rafaelsamenezes committed:
Merge branch 'master' into memcpy-optimization

rafaelsamenezes committed:
Merge branch 'master' into memcpy-optimization

rafaelsamenezes committed:
Handle infinite-sized array exceptions in memcpy intrinsic

- Added catch blocks for `array_type2t::inf_sized_array_excp` in both
occurrences of the `intrinsic_memcpy` function.
- Ensured that the function handles potential errors gracefully by
bumping the call and returning.

rafaelsamenezes committed:
Merge branch 'master' into memcpy-optimization

rafaelsamenezes committed:
Fix test failures

Increased the property count and pass/fail counts for the regression
tests. Specifically:

- Updated `regression/python/list-with-char_fail/test.desc` to reflect
50 verified properties (up from 44), with 48 passing and 2 failing.
- Updated `regression/python/union2_fail/test.desc` to reflect 63
verified properties (up from 62), with 59 passing and 4 failing.

rafaelsamenezes committed:
Merge branch 'master' into memcpy-optimization

rafaelsamenezes committed:
Merge branch 'master' into memcpy-optimization
